rfd/0134-cloud-first-deployment.md

```diff
@@ -0,0 +1,182 @@
+to "SaaSify" Teleport. The new (or rather, extended) Cloud-focused process aims
+to address shortcomings like:
+
+- Long release cycles where marquee new features get delivered to Cloud users
+  (roughly) once a quarter and are tied to major self-hosted releases.
+- Slow and monolithic release pipeline that always builds all artifacts and
+  does not allow for very fast Cloud rollouts.
+- Engineers being disconnected from Cloud deployments and unaware of when their
+  changes are getting rolled out.
+
+## Self-hosted release process (current)
+
+Before diving into the new process, let's set the stage with a high-level overview
+of the current release process.
+
+Current release versioning scheme follows semver, with quarterly major releases
+and monthly minor releases.
+
+Major releases are getting rolled onto Cloud quarterly, with a month's delay.
+Major features are reserved for major releases, which leads to Cloud customers
+getting major features with a month's delay.
+
+The release process is triggered on a Github tag creation and builds the full
+suite of artifacts that takes 2-3 hours so it's unrealistic to do multiple
+deploys a day.
+
+Cloud upgrade happens once a week to a release published the previous week, and
+engineers oftentimes don't have a clear idea of when their change is being
+deployed.
+
+## Cloud release process
+
+In the Cloud-first world, release that's running on Cloud does not necessarily
+have a corresponding self-hosted release tag. For this reason, we can't reuse
+the current versioning scheme and apply it to both Cloud and self-hosted.
+
+Instead, we want to get to a place where Cloud and self-hosted releases are
+decoupled, with Cloud releases being regularly cut from the main development
+branch and self-hosted releases cut from their respective release branches.
+
+Although the concept of semantic versioning loses most of its meaning with the
+continuous release train, we do need to keep it for Cloud releases to be able
+to use automatic upgrades system we built.
+
+With this in mind, the Cloud version will consist of:
+
+- Major: "development" version number on master branch (latest stable release + 1).
+- Minor/patch: always 0.
+- Prerelease: `-cloud.X` with a perpetually increasing "build" number which will
+  provide enough semver comparison guarantees for auto-upgrades to work.
+
+Once it is time for the next self-hosted major release (every 3-4 months), the
+new release branch will cut and the Cloud release version major component will
+increment.
+
+                                14.0.0     14.0.1     14.0.2
+                     branch/v14 |----------|----------|---------....---->
+                                |
+                                |
+            14.0.0-cloud.1      14.0.0-cloud.10    15.0.0-cloud.11    15.0.0-cloud.12    16.0.0-cloud.13
+master -----|-------....--------|------------------|-------....-------|------------------|--------------....---->
```

Comment:
(Minor) Did we talk about why cloud.X version doesn't reset when we increment the major version? E.g. why `15.0.0-cloud.11` rather than `15.0.0-cloud.0`?

Reply:
I don't think we achieved consensus on it but I agree resetting the counter makes more sense than keep forever incrementing it. Updated.